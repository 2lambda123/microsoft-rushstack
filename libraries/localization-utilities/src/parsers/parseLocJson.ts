// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { resolve } from 'path';

import { JsonFile, JsonSchema } from '@rushstack/node-core-library';

import { ILocalizationFile, IParseFileOptions } from '../interfaces';

const LOC_JSON_SCHEMA_PATH: string = resolve(__dirname, '../schemas/locJson.schema.json');
const LOC_JSON_SCHEMA: JsonSchema = JsonSchema.fromFile(LOC_JSON_SCHEMA_PATH);

/**
 * @public
 */
export function parseLocJson(options: IParseFileOptions): ILocalizationFile {
  const parsedFile: ILocalizationFile = JsonFile.parseString(options.content);
  try {
    LOC_JSON_SCHEMA.validateObject(parsedFile, options.filePath);
  } catch (e) {
    throw new Error(`The loc file is invalid. Error: ${e}`);
  }
  return parsedFile;
}
