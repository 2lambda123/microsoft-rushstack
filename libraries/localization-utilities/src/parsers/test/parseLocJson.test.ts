// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { parseLocJson } from '../parseLocJson';

describe(parseLocJson.name, () => {
  it('parses a valid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        comment: 'A string'
      }
    });

    expect(
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toMatchSnapshot();
  });

  it('throws on invalid file', () => {
    const content: string = JSON.stringify({
      foo: {
        value: 'Foo',
        baz: 'A string'
      }
    });

    expect(() =>
      parseLocJson({
        content,
        filePath: 'test.loc.json'
      })
    ).toThrowErrorMatchingSnapshot();
  });
});
