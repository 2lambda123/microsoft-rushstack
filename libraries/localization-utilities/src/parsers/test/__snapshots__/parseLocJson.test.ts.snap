// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLocJson parses a valid file 1`] = `
Object {
  "foo": Object {
    "comment": "A string",
    "value": "Foo",
  },
}
`;

exports[`parseLocJson throws on invalid file 1`] = `
"The loc file is invalid. Error: Error: JSON validation failed:
test.loc.json

Error: #/foo
       Additional properties not allowed: baz"
`;
